presto-main/src/main/java/com/facebook/presto/spiller/SpillSpaceTracker.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spiller;
+
+import com.facebook.presto.ExceededSpillLimitException;
+import com.facebook.presto.spi.QueryId;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.DataSize;
+import org.weakref.jmx.Managed;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import static com.facebook.presto.operator.Operator.NOT_BLOCKED;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.units.DataSize.succinctBytes;
+import static java.util.Objects.requireNonNull;
+
+public class SpillSpaceTracker
+{
+    private final long maxBytes;
+
+    @GuardedBy("this")
+    private long currentBytes;
+
+    @GuardedBy("this")
+    // TODO: It would be better if we just tracked QueryContexts, but their lifecycle is managed by a weak reference, so we can't do that
+    private final Map<QueryId, Long> queryMemoryReservations = new HashMap<>();
```

Comment:
Hmm, I copied it because MemoryPool uses it, but I see that this logic is duplicated in the QueryContext. I'll remove it, thanks.